### PR TITLE
Add support for building images from ostree containers and the Fedora CoreOS qcow2 image type

### PR DESCRIPTION
**Pull request by achilleas-k**
This PR adds support for the org.osbuild.ostree.deploy.container stage and a new image type that uses it called `coreos-qcow2` for Fedora.

I tested the new image type locally with the container we added to the gitlab registry for testing the sample manifest in osbuild.
Fedora CoreOS booted

Some notes and open questions:
- I use the same pipeline as the regular ostree deployment with a condition on the type of source.  I think it wasn't worth making a whole separate pipeline for this, but we can revisit this decision if the two start diverging.
- The new image type is added to Fedora which means that if this is merged and updated in osbuild-composer it will be available to build on-prem.  If we want this, we'll have to figure out the command line options for specifying a container source and potentially a target-imgref.
- When specifying a container source to deploy, you must specify the ref (otherwise the default will be assumed).  This requirement also exists for ostree commits from a repo, but I wonder: Is there a way to determine the container's ostree ref without downloading it?

You can generate the manifest for this image with:
```
go run ./cmd/gen-manifests --distros fedora-38 --arches x86_64 --images coreos-qcow2 --output .
```
or build it directly with
```
go build -o bin/build ./cmd/build
sudo ./bin/build --distro fedora-38 --image coreos-qcow2 --output builds --config ./test/configs/ostree-container-pull.json
```

You can also add a user to the config file which will create a user after the deployment before finalising the image:
```json
{
  "name": "ostree-container-pull",
  "ostree": {
    "ref": "ostree/1/1/0",
    "container": "registry.gitlab.com/redhat/services/products/image-builder/ci/images/fedora-coreos:stable",
    "tls-verify": true
  },
  "blueprint": {
    "customizations": {
      "user": [
        {
          "name": "<login>",
          "password": "<password>",
          "key": "<pubkey>"
        }
      ]
    }
  }
}
```

### CODE REVIEW

File: pkg/image/ostree_disk.go
Comment (bcl) [inline pkg/image/ostree_disk.go:45]: Why make this a pointer instead of using empty string to indicate it isn't set?

Comment (achilleas-k) [reply pkg/image/ostree_disk.go:45]: I don't really have a good answer.  I think I picked up a habit of making optionals `nil`-able but that's not a good reason.  You're right, an plain string with empty as the "off" value would work better and require less checks.

Comment (cgwalters) [reply pkg/image/ostree_disk.go:45]: mumble mumble `Option<String>`

File: pkg/distro/fedora/images.go
Comment (ondrejbudai) [inline pkg/distro/fedora/images.go:676]: Does `Keyboard` and `Locale` options have any effects on the manifest?

Comment (achilleas-k) [reply pkg/distro/fedora/images.go:676]: Adds the appropriate stages to the deployment pipeline: https://github.com/achilleas-k/images/blob/b19437c5d35c5c9028a37b14048352360288e79a/pkg/manifest/ostree_deployment.go#L361-L377

Comment (achilleas-k) [reply pkg/distro/fedora/images.go:676]: Those should actually be part of some image specific configuration (imageconfig or default workload or something) rather than hardcoded in the image function, but that's a more general problem I'm cleaning up separately.

File: pkg/image/ostree_disk.go
Comment (achilleas-k) [reply pkg/image/ostree_disk.go]: Done!

File: pkg/distro/fedora/images.go
Comment (ondrejbudai) [reply pkg/distro/fedora/images.go:676]: Makes me wonder: Should we meddle with the default keyboard/locale while deploying the container? I think I would rather leave the defaults as defined in the container.

Comment (achilleas-k) [reply pkg/distro/fedora/images.go]: Dropped.  But we should allow customizing these through the blueprint.  For all ostree images.  And the Anaconda ISO through kickstart (now that we added the commands in osbuild).

I think it's time to do the rewrite of the allowed options for image types.

Comment (ondrejbudai) [reply pkg/distro/fedora/images.go]: Agreed, for every image type, we should support as many customizations possible. :)

**Review by ondrejbudai — COMMENTED** [review]
I'm very happy with this PR and I haven't found any issues. However, it's a big one, so I would love a third pair of eyes to take a look. @achilleas-k should we ask someone explicitly? Any ideas whom? :)

File: pkg/distro/fedora/distro.go
Comment (teg) [inline pkg/distro/fedora/distro.go:710]: What about aarch64?

**Review by teg — DISMISSED** [review]
I think I noticed an oversight, but would rather fix that in a follow-up than rebasing it tbh, this is plenty useful as it is.

File: pkg/distro/fedora/distro.go
Comment (achilleas-k) [reply pkg/distro/fedora/distro.go:710]: I added it at first but then realised I needed to get the base container in our CI registry to test it and decided to get this open so people can review and I can add it in a follow-up.

File: pkg/manifest/ostree_deployment.go
Comment (dustymabe) [inline pkg/manifest/ostree_deployment.go]: We don't really need a pull stage at all anymore since https://github.com/osbuild/osbuild/pull/1393 - just specify the commit as an input to the deploy stage.

File: pkg/manifest/ostree_deployment.go
Comment (dustymabe) [inline pkg/manifest/ostree_deployment.go:244]: Is this stage always added? IIUC it's not always needed. For example, in Fedora we deliver remote configuration via the `fedora-repos-ostree` RPM:

```
[core@rpi4 ~]$ rpm -ql fedora-repos-ostree
/etc/ostree/remotes.d
/etc/ostree/remotes.d/fedora-compose.conf
/etc/ostree/remotes.d/fedora.conf
```

- [source](https://src.fedoraproject.org/rpms/fedora-repos/blob/rawhide/f/fedora.conf)

File: pkg/manifest/ostree_deployment.go
Comment (dustymabe) [inline pkg/manifest/ostree_deployment.go:253]: same for this one. we might not always want it.

File: pkg/image/ostree_disk.go
Comment (dustymabe) [inline pkg/image/ostree_disk.go:59]: to tamp down on confusion we could name this one `NewFromContainerOSTreeDiskImage()` and rename the other one to be `NewFromOSTreeCommitOSTreeDiskImage()`. The words "Container" and "Image" are both getting overloaded a lot and when I first read this I could have though we were creating a Container Image.

File: pkg/manifest/ostree_deployment.go
Comment (dustymabe) [inline pkg/manifest/ostree_deployment.go:199]: this looks very specific to a particular caller (Automotive? Edge?)

I think I would pull out `coreos.no_persist_ip` into something else. There is a new `kernel-cmdline.bls-append` stage in https://github.com/osbuild/osbuild/pull/1429

File: pkg/manifest/ostree_deployment.go
Comment (dustymabe) [inline pkg/manifest/ostree_deployment.go:308]: I suspect a lot of this is specific to IoT/Edge/Automotive

One thing that is interesting is that the `NewIgnitionStage` `Network:` stuff where `systemd.firstboot=off` and `systemd.condition-first-boot=true` get set is kind of an abuse of something that isn't really public API. It's meant to be an internal implementation detail.

File: pkg/ostree/ostree.go
Comment (dustymabe) [inline pkg/ostree/ostree.go:82]: since there are multiple things here that are "remote" artifacts it might be a little confusing to have this just apply to one of them and still name it `TLSVerify`.

One thing we could do here is break out sets of options that go together into subordinate objects.

i.e.

```
type ContainerImageOptions struct {
    Container string `json:"container"`
    TLSVerify *bool `json:"tls-verify"`
}
...
```

File: pkg/distro/fedora/images.go
Comment (dustymabe) [inline pkg/distro/fedora/images.go:714]: There is a lot going on here. Maybe @achilleas-k we can get together and step through some of it.

File: pkg/distro/fedora/distro.go
Comment (dustymabe) [inline pkg/distro/fedora/distro.go:387]: I probably need to dig into `iotBasePartitionTables` a little bit more to understand why they are being used here.

File: pkg/manifest/ostree_deployment.go
Comment (dustymabe) [inline pkg/manifest/ostree_deployment.go:344]: not sure either.. from my side this feels like something that would be done when composing the ostree commit or oscontainer rather than when composing an image. IOW I feel like this shouldn't be part of Image Builder at all?

File: pkg/distro/distro_test.go
Comment (dustymabe) [inline pkg/distro/distro_test.go:50]: is this supposed to be a container pullspec? those don't often include `https://` do they?

File: pkg/osbuild/ostree_deploy_container_stage.go
Comment (achilleas-k) [reply pkg/osbuild/ostree_deploy_container_stage.go:52]: Very much so.

File: pkg/manifest/ostree_deployment.go
Comment (achilleas-k) [reply pkg/manifest/ostree_deployment.go]: I want to keep the existing images unchanged while adding support for the new stuff to avoid any side effects or confusion.  I intend to drop the pull stage in a follow-up, but for now I'd like the existing images to change as little as possible and so I'm limiting their changes to just a stage reordering.

File: pkg/manifest/ostree_deployment.go
Comment (achilleas-k) [reply pkg/manifest/ostree_deployment.go:244]: We used to configure this before the package was created.  I have [a separate PR](https://github.com/osbuild/images/pull/95) that's been open for a while but configuring the remote through the package has been causing issues with our tests in osbuild-composer (https://github.com/osbuild/osbuild-composer/pull/3635).

I've been looking into it on and off and have no idea what the issue is, but I'd love to get your input on it.

File: pkg/manifest/ostree_deployment.go
Comment (achilleas-k) [reply pkg/manifest/ostree_deployment.go:253]: In what situation would we not need this?  We've always had it for deployments so I always considered it necessary.  The osbuild stage referenceshttps://github.com/coreos/fedora-coreos-config/blob/0b50084c9605d40c01c083e8b7b8efd9695d13a5/overlay.d/05core/usr/lib/dracut/modules.d/40ignition-ostree/ignition-ostree-populate-var.sh.

File: pkg/image/ostree_disk.go
Comment (achilleas-k) [reply pkg/image/ostree_disk.go:59]: That's a fair point.  The name of the pipeline in the manifest is `ostree-raw-image` for both and the type they return is `OSTreeDiskImage` so maybe that could be a prefix.
I think I'll go with
`NewOSTreeDiskImageFromCommit` and `NewOSTreeDiskImageFromContainer`.

File: pkg/manifest/ostree_deployment.go
Comment (achilleas-k) [reply pkg/manifest/ostree_deployment.go:199]: This was added by the edge/iot team as part of adding ignition support https://github.com/osbuild/osbuild-composer/pull/3130
@runcom: is this necessary for all ignition-enabled images?

File: pkg/manifest/ostree_deployment.go
Comment (achilleas-k) [reply pkg/manifest/ostree_deployment.go:308]: Again, I'll defer to @runcom.

File: pkg/ostree/ostree.go
Comment (achilleas-k) [reply pkg/ostree/ostree.go:82]: There's two things going on here that I don't like and I'm unsure of:
- The `ostree.ImageOptions` and more generally the [`distro.ImageOptions`](https://github.com/osbuild/images/blob/86c3f08097f0d409019f70cc0770fd4bcc564868/pkg/distro/distro.go#L132-L138) that exist separately from things like the blueprint customizations.  These I want to merge into a build config.
- The way that users will be specifying an ostree container for a build when this gets into osbuild-composer.

The way I added the options now sort of reflects things that might happen in composer, but that's just leftover internal structure from the split.

All that said, none of this is considered "stable API" (yet) in images, so I did the quickest thing that felt natural at the time.  You're right though, a completely separate object is nicer.  Getting the "build config" done soon would be even better.

File: pkg/distro/fedora/images.go
Comment (achilleas-k) [reply pkg/distro/fedora/images.go:714]: Sure, we can do that.  For a high level description though: Almost all of this function is essentially creating the `img`
```go
img := image.NewOSTreeContainerDiskImage(container, ref)
```
followed by setting options on the image based on image type settings and user customizations.

File: pkg/distro/fedora/distro.go
Comment (achilleas-k) [reply pkg/distro/fedora/distro.go:387]: I based the image off the iot-raw-image and reused that partition table.  I would say, instead of digging, point me to the default partition table you use in coreos images and I'll replicate it here.

File: pkg/manifest/ostree_deployment.go
Comment (dustymabe) [reply pkg/manifest/ostree_deployment.go]: ok

File: pkg/manifest/ostree_deployment.go
Comment (achilleas-k) [reply pkg/manifest/ostree_deployment.go:344]: Locking the root password during deployment is something that's done for the official Fedora IoT image: https://pagure.io/fedora-kickstarts/blob/main/f/fedora-iot.ks#_9

We were replicating that here.

File: pkg/manifest/ostree_deployment.go
Comment (dustymabe) [reply pkg/manifest/ostree_deployment.go:244]: looks like https://github.com/ostreedev/ostree/discussions/3079#discussioncomment-7339766 seems to have figured it all out?

File: pkg/distro/distro_test.go
Comment (achilleas-k) [reply pkg/distro/distro_test.go:50]: It could be any string since the test doesn't really do anything with it, but you're right.  Might as well make it look like a real spec.

File: pkg/manifest/ostree_deployment.go
Comment (dustymabe) [reply pkg/manifest/ostree_deployment.go:253]: well, for Fedora/RH CoreOS for instance, we have the file you linked that runs on boot and does those things, so we don't need them in the disk image.

File: pkg/image/ostree_disk.go
Comment (dustymabe) [reply pkg/image/ostree_disk.go:59]: I like those even better.

File: pkg/manifest/ostree_deployment.go
Comment (dustymabe) [reply pkg/manifest/ostree_deployment.go:199]: I can say at least for CoreOS, we don't want it :)

File: pkg/ostree/ostree.go
Comment (dustymabe) [reply pkg/ostree/ostree.go:82]: I'm OK with whatever.

File: pkg/manifest/ostree_deployment.go
Comment (achilleas-k) [reply pkg/manifest/ostree_deployment.go:244]: It seems to be something else.  I should write it up somewhere else but here's how the problem manifests:
- Build an image with the ostree remote stage (like we do now).
- Boot it
- Remove the remote
- Add a new remote with the same name, but with a different URL
- Add an upgrade commit to the location of the new remote
- Run `rpm-ostree upgrade`

This works.  The update gets pulled and all is well.

If I build an image without the ostree remote stage but with the packaged remote configuration and perform the same steps, then `rpm-ostree upgrade` doesn't do anything.  I can `ostree remote log` and `ostree remote summary` etc, I can see the upgrade commit in the repo.  I can even `rpm-ostree rebase` onto it.  But `rpm-ostree upgrade` refuses to pick it up.

File: pkg/distro/fedora/distro.go
Comment (dustymabe) [reply pkg/distro/fedora/distro.go:387]: [this](https://github.com/osbuild/osbuild/blob/5c345fb3fa5d9fa5804ba74e7dc4445ae8ff1010/test/data/manifests/fedora-coreos-container.mpp.yaml#L2-L27) is what I've been using for now.

There is one theme here, though, that I think we keep bumping up into, which is where definitions like this should live. I think it was also touched on in https://github.com/osbuild/osbuild-deploy-container/issues/4 
      I

Comment (achilleas-k) [reply pkg/distro/fedora/distro.go:387]: For Image Builder (osbuild-composer and the service), the defaults are defined here.  For all the RHEL images we build for example as part of the composes in Brew and for all the images in c.rh.c/ib, this is the source of truth right now.

File: pkg/manifest/ostree_deployment.go
Comment (dustymabe) [reply pkg/manifest/ostree_deployment.go:244]: dumb question.. there is a new commit in the history for that ref to upgrade to?

Comment (achilleas-k) [reply pkg/manifest/ostree_deployment.go:244]: Yes, I create the repo, serve it over a web server, and pull an upgrade commit into it in both scenarios.

I'd love to show you the whole thing.  I'll tidy up my repro scripts and put them somewhere tomorrow as well, show them around to anyone who might have some idea about what's going on.

Comment (achilleas-k) [reply pkg/manifest/ostree_deployment.go:244]: Solved this issue after 5 mins of showing Dusty the issue.  I wasn't defining a remote name when deploying.  Fixed separately.  Thanks @dustymabe!!

File: pkg/manifest/ostree_deployment.go
Comment (achilleas-k) [reply pkg/manifest/ostree_deployment.go:253]: This stage isn't run when building the CoreOS image.  Resolving.

**Comment by achilleas-k** [comment]
Partially addressed comments (and rebased on `main`).  Will resume addressing comments later.

File: pkg/osbuild/ostree_deploy_container_stage.go
Comment (achilleas-k) [reply pkg/osbuild/ostree_deploy_container_stage.go:8]: This is a fair point, but I also like that the regex string is a `const`.  I don't expect anything will modify the regex at runtime, but it's nice to have that guarantee.

File: pkg/manifest/ostree_deployment.go
Comment (achilleas-k) [reply pkg/manifest/ostree_deployment.go:199]: Fixed in #272.

File: pkg/image/ostree_disk.go
Comment (achilleas-k) [reply pkg/image/ostree_disk.go:69]: Fixed in #272.

**Comment by achilleas-k** [comment]
Moved most of the commits to #272.  I'll rebase this PR so that it only adds the coreos image once that's merged.

File: pkg/manifest/ostree_deployment.go
Comment (runcom) [reply pkg/manifest/ostree_deployment.go:305]: So, I think we can’t really say that’s not public api as it has been used in the past by coreos too (where we’ve learned about it) and even if undocumented it acts as an api to me 🤷🏻‍♂️ anyhow, I agree it’s a huge ack and it’s meant as a workaround that we need to solve (I’ll prioritize investigation and fix soon)

**Comment by supakeen** [comment]
@achilleas-k should be able to rebase now as #272 has landed \o/

**Comment by achilleas-k** [comment]
Will rebase and fix conflicts then start addressing the remaining comments.

**Pull request closed without merging**
